### PR TITLE
KT-16260 Add intention to specify all types explicitly in destructuring assignment

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-16260

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        if (element.containingFile is KtCodeFragment) return null
+        val entries = element.noTypeReferenceEntries()
+        if (entries.isEmpty()) return null
+        if (entries.any { SpecifyTypeExplicitlyIntention.getTypeForDeclaration(it).isError }) return null
+        return TextRange(element.startOffset, element.initializer?.let { it.startOffset - 1 } ?: element.endOffset)
+    }
+
+    override fun applyTo(element: KtDestructuringDeclaration, editor: Editor?) {
+        val entries = element.noTypeReferenceEntries()
+        if (editor != null && element.initializer != null)
```
Comment (yole) [inline idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt]: This behaves differently for a destructuring assignment and a destructuring used in a loop:

```
        val map = mapOf(1 to "two")
        for ((key, value) in map) { 
        }
```

The behavior should be the same in both cases.

File: idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import org.jetbrains.kotlin.types.isError
+
+class SpecifyTypeExplicitlyInDestructuringAssignmentIntention : SelfTargetingRangeIntention<KtDestructuringDeclaration>(
+        KtDestructuringDeclaration::class.java, "Specify all types explicitly in destructuring assignment"
+), LowPriorityAction {
+
+    override fun applicabilityRange(element: KtDestructuringDeclaration): TextRange? {
+        if (element.containingFile is KtCodeFragment) return null
+        val entries = element.noTypeReferenceEntries()
+        if (entries.isEmpty()) return null
```
Comment (yole) [inline idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt]: This is redundant; the next line will also return if `entries` is empty.

File: idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt (hunk 1 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt]: If `&& element.initializer != null` is deleted, the behavior will be same:
```
val map = mapOf(1 to "two")
for ((key, value)<caret> in map) {
}
```

However, destructuring in lambda will not work correctly:
```
val map = mapOf(1 to "two")
map.forEach { (key, value)<caret> ->
}
```
I think the cause is `typeRef.isValid` returns false:
https://github.com/t-kameyama/kotlin/blob/827c57d6ed7492b5f41fbc4ac87b196deebde7bf/idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyIntention.kt#L159
I also confirmed that `typeRef.text` is `Any`.

Is there a way for `typeRef.isValid` to become `true`?

File: idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt (hunk 2 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt]: This is not redundant. `entries.any {}` returns false if `entries` is empty.

Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt]: Fixed

File: idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt (hunk 1 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt]: Fixed

**Comment by yole** [comment]
Merged manually (with some minor wording changes). Thanks a lot for your contribution!

**Pull request closed without merging**